Handle rlig for latin script

Connum opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This handles the `rlig` feature for latin scripts.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Despite being enabled for latin scripts even by default, `rlig` ligatures were only handled for Arabic script.

I did some minor refactoring of the ligature application to diminish code duplication betweeen arabic and latin script ligatures and the new rlig implementation. They now use the same method in a new `src/features/commonFeatureUtils.js` file, without breaking the existing file/module structure, with parameters for the script, feature and range. This might come in handy as we progress to implement new font features. Speaking of which, we should probably overhaul feature handling as we do so. This PR only enables `rlig` for the latin script, which fixes #554 (and fixes #479), while other scripts might need it as well and there are other features that might not be script-specific.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested with the Pecita font provided in #479 and added corresponding tests for it (the font is under SIL Open Font License so I was able to use it directly for the tests), not breaking any other tests.

## Screenshots (if appropriate):
before:
![image](https://user-images.githubusercontent.com/13076806/218817963-b13eca7c-96ed-41a3-96b7-a29a41b16dd8.png)
after (note the "qu" and "ck" shapes):
![image](https://user-images.githubusercontent.com/13076806/218817815-5f4a4ebc-279a-4aff-bbd9-08f13eb76255.png)
The rest of the differences to how it should look according to #479 can be attributed to not supporting the `calt` feature.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I did `npm run test` and all tests passed green (including code styling checks).
- [X] I have added tests to cover my changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the **README** accordingly.
- [X] I have read the **CONTRIBUTING** document.

rafallyczkowskiadylic committed:
feat: additional GPOS lookups parsers

- added GPOS LookupType 4 parser for MarkBasePosFormat1 support
- added GPOS LookupType 9 parser for Extension Positioning SubtableFormat1 support
- tests coverage for all supported and not supported lookups of GPOS

TODO: Add a support device offsets for anchor (Parser.anchor) table format 3.

rafallyczkowskiadylic committed:
feat: introduced glyphs positioning features (kern, mark)

- introduced an union, ordered features lookup list
- refactored the Font options features update method
- supported kerning table
- supported kern lookup table
- supported mark to base attachments

rafallyczkowskiadylic committed:
fix: code style fixes.

rafallyczkowskiadylic committed:
fix: kerning pos fix for pair

rafallyczkowskiadylic committed:
fix: kern processing value to support entire sequence

- fix for kern sequence
- tests coverage

Connum committed:
refactored ligature feature application to diminish duplicate code

Connum committed:
apply rlig feature for latin script

Connum committed:
add tests for rlig feature enabled/disabled for latin script

Connum committed:
force rlig feature to true in index.html demo

ILOVEPIE commented:
This might conflict with #535 and #557. Which should we merge first, one of those two PRs reworks the feature system (can't remember which one).

ILOVEPIE commented:
@Connum do we want to merge this first or rework it once #535 and #557 are merged?

Connum commented:
Good question... Both of them made changes to bidi.js which I also altered, and both of them could probably benefit from the common helper function that I introduced. I think it's easier to merge those two first and have me adapt the rlig code to those changes instead of heaving to incorporate mine into both of the PRs.

Connum committed:
Merge branch 'master' into 554-latin-rlig

# Conflicts:
#	src/bidi.js
#	src/features/arab/arabicRequiredLigatures.js
#	src/features/latn/latinLigatures.js

Connum committed:
fix broken import after merge

Connum commented:
up-to-date with current master, blocked by #535 and #557

Connum committed:
fix test font file path

Connum committed:
did I really forget to commit the test font and no-one noticed? ;)

Connum committed:
Merge branch 'master' into 554-latin-rlig

Connum commented on src/features/commonFeatureUtils.js:
```diff
@@ -0,0 +1,37 @@
+import { ContextParams } from '../tokenizer';
```
I just noticed that the file extensions are missing, but are now required since we changed the build process and the library may be used in module form on the web, where we need the file extensions. I'm fixing it right now.

rafallyczkowskiadylic committed:
Merge branch 'master' into feature/mtbase-merge-testing

# Conflicts:
#	src/font.js
#	test/layout.js

rafallyczkowskiadylic committed:
Merge branch 'master' into feature/mark-to-base-attachment-positioning

rafallyczkowskiadylic committed:
fix: pointer offset fix, typos

rafallyczkowskiadylic committed:
Merge branch 'master' into feature/mark-to-base-attachment-positioning

# Conflicts:
#	src/font.js

rafallyczkowskiadylic committed:
fix: tests after merge

rafallyczkowskiadylic committed:
Merge branch 'master' into feature/mark-to-base-attachment-positioning

rafallyczkowskiadylic committed:
fix: fixed lint code after merge

rafallyczkowskiadylic committed:
fix: Fixes the substitution algorithm and support for multiSubstitutionFormat1

- substitution algorithm to match specficiation (lookup - apply - lookup - apply)
- added support for multiSubstitutionFormat1 substitutions
- updated applySubstitution method for more flexibility
- tests coverage

rafallyczkowskiadylic committed:
fix: context condition check

rafallyczkowskiadylic committed:
fix: code linting and comments

rafallyczkowskiadylic committed:
Merge pull request #5 from Adylic/bugfix/substitution-algorithm-fix

Bugfix/substitution algorithm fix

Connum committed:
Merge branch 'master' into 554-latin-rlig

# Conflicts:
#	docs/index.html
#	src/bidi.js
#	src/features/arab/arabicRequiredLigatures.js
#	src/features/latn/latinLigatures.js

Connum committed:
fix cmap table selection

Connum committed:
implement reading, writing and drawing of CFF PaintType and StrokeWidth

Connum committed:
clarify test font license

Connum committed:
correct 2x2 glyf transformation (fix #432)

Connum committed:
better browser/node detection

Connum committed:
add eslint-import rule no-cycle

Connum committed:
Merge branch 'master' into fix/459

Connum committed:
Merge branch 'master' into fix/649

Connum committed:
Merge branch 'master' into fix/135

Connum committed:
Merge branch 'master' into fix/174

Connum committed:
Merge branch 'master' into dev/eslint-import-no-cycle

Connum committed:
Update page.yml

Connum committed:
Update ci.yml

Connum committed:
Merge pull request #1 from opentypejs/dev/eslint-import-no-cycle

Dev/eslint import no cycle

Connum committed:
better browser/node detection #653

Fix/174

Connum committed:
Merge remote-tracking branch 'upstream/pr/652' into next/main

Connum committed:
Merge remote-tracking branch 'upstream/pr/651' into next/main

Connum committed:
Merge remote-tracking branch 'upstream/pr/650' into next/main

Connum committed:
add unicode-report workflow

Connum committed:
Create report-counter.js

Connum committed:
Add files via upload

Connum committed:
Update unicode-report.yml

Connum committed:
Merge branch 'main' of https://github.com/Connum/opentype.js-next into next/main

# Conflicts:
#	.github/workflows/scripts/report-counter.js
#	.github/workflows/unicode-report.yml

Connum committed:
Merge branch 'feature/mark-to-base-attachment-positioning' of https://github.com/Adylic/opentype.js into Adylic-feature/mark-to-base-attachment-positioning

# Conflicts:
#	src/font.js

Connum committed:
Update unicode-report.yml to allow manual dispatch

Connum committed:
Update unicode-report.yml

Connum committed:
Update unicode-report.yml

Connum committed:
Update unicode-report.yml

Connum committed:
Update unicode-report.yml

Connum committed:
Update unicode-report.yml

Connum committed:
Update unicode-report.yml

Connum committed:
Update unicode-report.yml

Connum committed:
Update unicode-report.yml

Connum committed:
Merge branch 'main' into 554-latin-rlig

Connum commented:
I somehow managed to mess up this branch... Going to convert to draft until I find time to fix this.

ILOVEPIE commented:
@Connum can we get this updated?
